zarr/store/dictstore.py

```diff
@@ -0,0 +1,85 @@
+
+import numpy as np
+from zarr.store.base import ArrayStore
+from zarr.util import normalize_cparams, normalize_shape, normalize_chunks, \
+    normalize_resize_args
+
+class DictStore(object):
+    """ ArrayStore implemented from three MutableMappings
+
+    Parameters
+    ----------
+    data: MutableMapping
+        Holds bulk data
+    attrs: MutableMapping
+        Holds user defined attributes
+    meta: MutableMapping
+        Holds array metadata, like shape, chunks, dtype, etc..
+    **extra_meta: Keyword arguments
+        shape, chunks, dtype, etc..  See MemoryStore for options
+
+    Examples
+    --------
+    >>> data = {}
+    >>> attrs = {}
+    >>> meta = {}
+    >>> store = DictStore(data, attrs, meta, shape=(1000,), chunks=(100,))
+    >>> meta['shape']
+    (1000,)
+
+    >>> import zarr
+    >>> x = zarr.Array(store)
+    >>> x[:] = 1
+    >>> sorted(data.keys())
+    ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']
+    """
+    def __init__(self, data, attrs, meta, count_cbytes=False, **extra_meta):
+        meta.update(extra_meta)
+
+        # normalize arguments
+        meta['cname'], meta['clevel'], meta['shuffle'] = \
+                normalize_cparams(meta.get('cname', None),
+                                  meta.get('clevel', None),
+                                  meta.get('shuffle', None))
+        meta['shape'] = normalize_shape(meta['shape'])
+        meta['chunks'] = normalize_chunks(meta.get('chunks', None),
+                                          meta['shape'])
+        meta['dtype'] = np.dtype(meta.get('dtype', None))  # float64 default
+        meta['fill_value'] = meta.get('fill_value', None)
+
+        self.data = data
+        self.attrs = attrs
+        self.meta = meta
+        self.count_cbytes = count_cbytes
```

Comment:
OK by me.

On Fri, Apr 22, 2016 at 3:35 PM, Matthew Rocklin <notifications@github.com>
wrote:

> In zarr/store/dictstore.py
> <https://github.com/alimanfoo/zarr/pull/23#discussion_r60748687>:
>
> > +
> > +        # normalize arguments
> > +        meta['cname'], meta['clevel'], meta['shuffle'] = \
> > +                normalize_cparams(meta.get('cname', None),
> > +                                  meta.get('clevel', None),
> > +                                  meta.get('shuffle', None))
> > +        meta['shape'] = normalize_shape(meta['shape'])
> > +        meta['chunks'] = normalize_chunks(meta.get('chunks', None),
> > +                                          meta['shape'])
> > +        meta['dtype'] = np.dtype(meta.get('dtype', None))  # float64 default
> > +        meta['fill_value'] = meta.get('fill_value', None)
> > +
> > +        self.data = data
> > +        self.attrs = attrs
> > +        self.meta = meta
> > +        self.count_cbytes = count_cbytes
>
> I opted to avoid the properties and private data here. Directly adding the
> data to the class seemed simpler. Happy to revert if desired.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/alimanfoo/zarr/pull/23/files/62a172f8d3e2e7e2c590accffc3a9b558bc46f88#r60748687>
>



-- 
Alistair Miles
Head of Epidemiological Informatics
Centre for Genomics and Global Health <http://cggh.org>
The Wellcome Trust Centre for Human Genetics
Roosevelt Drive
Oxford
OX3 7BN
United Kingdom
Email: alimanfoo@googlemail.com <alimanfoo@gmail.com>
Web: http://purl.org/net/aliman
Twitter: https://twitter.com/alimanfoo
Tel: +44 (0)1865 287721